### PR TITLE
Verify participant onboarding status

### PR DESCRIPTION
**Pull request by cpeeyush**
## What this PR changes/adds

- As the registration service now has an endpoint to fetch the record of a participant, this can be utilized during participant onboarding to verify that onboarding completed successfully or not.
- Update to system-tests/README.md necessary for a successful build of Registration service
- Add health check to `system-tests/resources/cli-tools/Dockerfile` to have an indicator when container `cli-tools` is finished.
- Updated to use latest version of registration service.

## Why it does that

Validate if a participant onboarded successfully or not.

## Linked Issue(s)

Relates to https://github.com/agera-edc/MinimumViableDataspace/issues/46

## Checklist

- [ ] added appropriate tests?
- [x] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly?

### CODE REVIEW

File: .github/workflows/deploy.yaml [hunk 1]
```diff
@@ -462,6 +462,40 @@ jobs:
           REGISTRATION_SERVICE_VERSION: 0.0.1-SNAPSHOT
           DATASPACE_DID_HOST: ${{ needs.Deploy-Dataspace.outputs.dataspace_did_host }}
 
+      - name: 'Wait for participant registration'
+        run: |               
+          retryCount=0
+          maxRetryCount=30
+          onboardingCompleted=false
+          
+          while [ $retryCount -lt $maxRetryCount ]; do
```
Comment (marcgs) [inline .github/workflows/deploy.yaml]: Code seems to be (at least to some extent) duplicated from `entrypoint.sh` consider extracting into an own file and reusing.

File: system-tests/docker-compose.yml [hunk 2]
```diff
@@ -171,6 +171,7 @@ services:
       context: ./resources/cli-tools
     volumes:
       - ./resources:/resources
+    tty: true
```
Comment (marcgs) [inline system-tests/docker-compose.yml]: Not sure to understand why that is needed

File: system-tests/resources/cli-tools/Dockerfile [hunk 3]
```diff
@@ -15,4 +15,7 @@ COPY ./identity-hub-cli.jar .
 
 COPY ./entrypoint.sh .
 
+# health status is determined by the success of the entrypoint script
+HEALTHCHECK --start-period=60s --interval=5s CMD test -f finished.flag && exit 0 || exit 1
```
Comment (marcgs) [inline system-tests/resources/cli-tools/Dockerfile]: Setting "retries" would make this part more explicit

File: system-tests/resources/cli-tools/entrypoint.sh [hunk 4]
```diff
@@ -3,34 +3,92 @@
...
+  # Wait for participant registration.
+  waitForOnboarding "$participantName" "$participantDid"
 done
+
+# flag for healthcheck by Docker
+echo "finished" > finished.flag
+echo "Finished successfully! Keep the container running."
+
+# keep the container running
+sleep infinity
```
Comment (marcgs) [inline system-tests/resources/cli-tools/entrypoint.sh:78]: Why is this needed now and not before this change?

File: .github/workflows/deploy.yaml [hunk 5]
```diff
@@ -462,6 +462,40 @@ jobs:
           REGISTRATION_SERVICE_VERSION: 0.0.1-SNAPSHOT
           DATASPACE_DID_HOST: ${{ needs.Deploy-Dataspace.outputs.dataspace_did_host }}
 
+      - name: 'Wait for participant registration'
+        run: |               
+          retryCount=0
```
Comment (algattik) [inline .github/workflows/deploy.yaml]: why is this script duplicated (with entrypoint.sh) rather than extracted?

File: system-tests/README.md [hunk 6]
```diff
@@ -145,7 +145,8 @@ Copy Identity Hub client-cli jar which should be located at `<Identity-Hub-root-
 From the `Registration Service` root folder, execute the following command:
 
 ```bash
-./gradlew :launcher:shadowJar
+./gradlew -DuseFsVault="true" :launcher:shadowJar
```
Comment (algattik) [inline system-tests/README.md]:
```suggestion
./gradlew -DuseFsVault="true" shadowJar
```
and remove line below

File: system-tests/resources/cli-tools/Dockerfile (hunk 3 above)
Comment (algattik) [inline system-tests/resources/cli-tools/Dockerfile]:
```suggestion
HEALTHCHECK --start-period=60s --interval=5s CMD test -f finished.flag
```

File: system-tests/resources/cli-tools/entrypoint.sh [hunk 7]
```diff
@@ -3,34 +3,92 @@
...
-                -c="$participantDid" \
-                 participants add
+  # seed  vc for participant
+  seedVerifiedCredentials "$participantName" "$participantDid" "$region"
+
+  # Register dataspace participants
+  registerParticipant "$participantName" "$participantDid"
+
+  # Wait for participant registration.
+  waitForOnboarding "$participantName" "$participantDid"
```
Comment (algattik) [inline system-tests/resources/cli-tools/entrypoint.sh]: seems inefficient to wait in a loop like that. I'm open to a discussion on this. What's the time impact of this?

Would an alternative be to (1) rearrange the script to trigger all onboardings first and then wait for all to complete, or (2) run this container three times?

**Review by algattik — CHANGES_REQUESTED** [review]

File: system-tests/docker-compose.yml (hunk 2 above)
Comment (pkirch) [reply system-tests/docker-compose.yml:174]: That keeps the container running even the entrypoint has finished. It is useful to have a container to connect to for development and testing.

File: system-tests/resources/cli-tools/entrypoint.sh (hunk 4 above)
Comment (pkirch) [reply system-tests/resources/cli-tools/entrypoint.sh:78]: For local development it turned out to be useful to have this container keep running and being able to connect to.

File: system-tests/docker-compose.yml (hunk 2 above)
Comment (pkirch) [reply system-tests/docker-compose.yml]: Indeed, a left-over which we have introduced before we changed entrypoint.sh to keep running with the sleep command.

Can be removed.

File: system-tests/resources/cli-tools/Dockerfile (hunk 3 above)
Comment (pkirch) [reply system-tests/resources/cli-tools/Dockerfile]: I'll set the default value for `retries` explicitly.

File: system-tests/resources/cli-tools/entrypoint.sh (hunk 4 above)
Comment (algattik) [reply system-tests/resources/cli-tools/entrypoint.sh:78]: how does the user know when they can start the test then? before it was when the container exited

BTW could we make sure (in this or next PR) this point is made clear in https://github.com/agera-edc/MinimumViableDataspace/blob/main/system-tests/README.md

File: system-tests/resources/cli-tools/entrypoint.sh [hunk 8]
```diff
@@ -3,34 +3,92 @@
...
 done
+
+# flag for healthcheck by Docker
+echo "finished" > finished.flag
+echo "Finished successfully! Keep the container running."
```
Comment (algattik) [inline system-tests/resources/cli-tools/entrypoint.sh:75]: according to the docker-compose doc `--wait                      Wait for services to be running|healthy. Implies detached mode.`. This fix seems to be assuming the command waits for services to be healthy?

File: system-tests/docker-compose.yml (hunk 2 above)
Comment (chrislomonico) [inline system-tests/docker-compose.yml]: ```suggestion```
Should be removed unless there is a reason for the interactive terminal

**Review by chrislomonico — CHANGES_REQUESTED** [review]
please review comments

File: system-tests/docker-compose.yml (hunk 2 above)
Comment (pkirch) [reply system-tests/docker-compose.yml]: Already addressed in Marc's comment. Has been removed.

File: system-tests/resources/cli-tools/entrypoint.sh (hunk 7 above)
Comment (pkirch) [reply system-tests/resources/cli-tools/entrypoint.sh]: As discussed with @algattik, going for option 1 and separated seeding/registration from waiting.

File: system-tests/resources/cli-tools/Dockerfile (hunk 3 above)
Comment (pkirch) [reply system-tests/resources/cli-tools/Dockerfile]: Good catch! Simplified the health check as suggested.

File: system-tests/README.md (hunk 6 above)
Comment (pkirch) [reply system-tests/README.md]: Thanks! Changed and removed.

File: system-tests/resources/cli-tools/entrypoint.sh (hunk 4 above)
Comment (pkirch) [reply system-tests/resources/cli-tools/entrypoint.sh:78]: As discussed with @algattik updated the README to highlight this (changed) behavior.

File: .github/workflows/deploy.yaml (hunk 1 above)
Comment (pkirch) [reply .github/workflows/deploy.yaml]: I have consolidated the duplicated code.

File: .github/workflows/deploy.yaml (hunk 5 above)
Comment (pkirch) [reply .github/workflows/deploy.yaml]: I have consolidated the duplicated code.

**Review by marcgs — APPROVED** [review]

File: .github/workflows/deploy.yaml [hunk 9]
```diff
@@ -462,6 +462,24 @@ jobs:
...
+          hostDid="did:web:$DATASPACE_DID_HOST"
+          participantDid="did:web:$PARTICIPANT_DID_HOST"
+          
+          cmd="java -jar registration-service-cli-$REGISTRATION_SERVICE_VERSION-all.jar \
+                -d=$hostDid \
+                -k=participantkey.pem \
+                -c=$participantDid \
+                participants get"
+          
+          $GITHUB_WORKSPACE/system-tests/resources/cli-tools/validate_onboarding.sh "$participantDid" "$cmd"
```
Comment (algattik) [inline .github/workflows/deploy.yaml:474]:
```suggestion
          system-tests/resources/cli-tools/validate_onboarding.sh "$participantDid" "$cmd"
```

and 

```
        working-directory: .
```

(would be a nit, but since this pattern is used at line 409, it seems good to standardise)

File: .github/workflows/deploy.yaml [hunk 10]
```diff
@@ -462,6 +462,24 @@ jobs:
           REGISTRATION_SERVICE_VERSION: 0.0.1-SNAPSHOT
           DATASPACE_DID_HOST: ${{ needs.Deploy-Dataspace.outputs.dataspace_did_host }}
 
+      - name: 'Wait for participant registration'
+        run: |
+          
+          hostDid="did:web:$DATASPACE_DID_HOST"
```
Comment (algattik) [inline .github/workflows/deploy.yaml]: `host` is somewhat unexpected here, since the other variable is also a host.

Consider inlining those two

File: README.md [hunk 11]
```diff
@@ -81,6 +81,26 @@ standard scenario which can be optionally used with the local development enviro
 > services on the console set `EDC_CATALOG_CACHE_EXECUTION_PERIOD_SECONDS` to a higher value, e.g. 60, for each EDC
 > participant in `system-tests/docker-compose.yml`.
 
+> Note: The container `cli-tools` will turn into the state `healthy` after registering successfully all participants and
+> will keep running as an entrypoint to the services created by Docker compose. This is useful for local development in order
+> to manually check commands against the participants (e.g. `company1`, `company2`, `company3`).
```
Comment (algattik) [inline README.md:86]: great comment, thanks!

File: README.md [hunk 12]
```diff
@@ -81,6 +81,26 @@ standard scenario which can be optionally used with the local development enviro
...
+
+Sample how to enter the container `cli-tools` and test a command manually.
+
+```powershell
+PS C:\wb\edc\agera\mvd> docker exec -it cli-tools bash
+root@22345bf0c595:/app# java -jar registration-service-cli.jar \
```
Comment (algattik) [inline README.md]: please remove incidental strings (`wb`, `22345bf0c595`)

File: README.md [hunk 13]
```diff
@@ -81,6 +81,26 @@ standard scenario which can be optionally used with the local development enviro
...
+>                       -d=did:web:did-server:registration-service \
+>                       --http-scheme \
+>                       -k=/resources/vault/company1/private-key.pem \
+>                       -c=did:web:did-server:company1 \
+>                       participants get
+{
+  "did" : "did:web:did-server:company1",
+  "status" : "ONBOARDED"
+}root@22345bf0c595:/app#
```
Comment (algattik) [inline README.md]: remove prompt

File: system-tests/README.md [hunk 14]
```diff
@@ -178,7 +178,17 @@ Once completed, following services will start within their docker containers:
...
 
-_Note, the `Newman` docker container will automatically stop after seeding initial data from postman scripts and `cli-tools` container will also automatically stop after registering participants._
+> Note, the `Newman` docker container will automatically stop after seeding initial data from postman scripts.
+
+> The container `cli-tools` will turn into the state `healthy` after registering successfully all participants.
+
+Sample for confirming successful run of container `cli-tools`.
+
+```powershell
+PS C:\> docker ps
```
Comment (algattik) [inline system-tests/README.md]:
```suggestion
docker ps
```

Comment (algattik) [reply system-tests/README.md]: consider removing prompts from commands throughout the document, to make this clearer and more cross-shell.

File: system-tests/README.md [hunk 15]
```diff
@@ -178,7 +178,17 @@ Once completed, following services will start within their docker containers:
...
+CONTAINER ID   IMAGE                                     COMMAND                   CREATED              STATUS                        PORTS                                                                              NAMES
```
Comment (algattik) [inline system-tests/README.md:196]: consider moving command outputs to separate ``` blocks, to make it clearer and avoid "random" syntax highlighting. E.g. Typora gives strange colours here

image

File: system-tests/resources/cli-tools/entrypoint.sh [hunk 16]
```diff
@@ -3,34 +3,76 @@
...
+                 -i="did:web:did-server:gaia-x" \
+                 -k="/resources/vault/gaia-x/private-key.pem"
+   done
+}
+
+PARTICIPANTS=(company1:eu company2:eu company3:us)
+
+# Seed VCs and register participants.
+for participant in "${PARTICIPANTS[@]}"; do
+  participantArray=(${participant//:/ })
```
Comment (algattik) [inline system-tests/resources/cli-tools/entrypoint.sh:42]:
```suggestion
  participantArray=(${participant//:/})
```
nit

File: system-tests/resources/cli-tools/entrypoint.sh [hunk 17]
```diff
@@ -3,34 +3,76 @@
...
-                --http-scheme \
-                -k=/resources/vault/$participantName/private-key.pem \
-                -c="$participantDid" \
-                 participants add
+                 -i="did:web:did-server:gaia-x" \
+                 -k="/resources/vault/gaia-x/private-key.pem"
+   done
+}
+
+PARTICIPANTS=(company1:eu company2:eu company3:us)
```
Comment (algattik) [inline system-tests/resources/cli-tools/entrypoint.sh:38]: shouldn't this be pulled from `participants.json`?

File: system-tests/resources/cli-tools/validate_onboarding.sh [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    status=$($cmd|jq ".status")
+
+    echo "Status is $status"
+
+    if [ "$status" == "\"ONBOARDED\"" ]; then
+        echo "$participantDid is onboarded successfully"
+        onboardingCompleted=true
+        break
+    else
+        echo "Onboarding is not completed yet for $participantDid. Waiting for 1 seconds..."
```
Comment (algattik) [inline system-tests/resources/cli-tools/validate_onboarding.sh]:
```suggestion
        echo "Onboarding is not completed yet for $participantDid. Waiting for 1 second..."
```

File: system-tests/README.md (hunk 14 above)
Comment (algattik) [inline system-tests/README.md]:
```suggestion
docker ps -a
```

Otherwise if cli-tools fails, its entry is not shown. With `-a`:

```
25207e27673c   system-tests_cli-tools                    "/bin/sh -c \"/app/en…"   2 minutes ago   Exited (1) 2 minutes ago                                                                                           cli-tools
```

**Review by algattik — APPROVED** [review]

File: system-tests/README.md (hunk 15 above)
Comment (chrislomonico) [reply system-tests/README.md:196]: this is a great approach and helps the commands to be clear and stand out in the `docs`

File: system-tests/resources/cli-tools/validate_onboarding.sh [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+retryCount=0
+maxRetryCount=30
+onboardingCompleted=false
+
+while [ $retryCount -lt $maxRetryCount ]; do
+
+    status=$($cmd|jq ".status")
+
+    echo "Status is $status"
```
Comment (chrislomonico) [inline system-tests/resources/cli-tools/validate_onboarding.sh]: nit:
```suggestion
    echo "Status: $status"
```

**Review by chrislomonico — APPROVED** [review]
approved pending resolution of open comments

File: .github/workflows/deploy.yaml (hunk 9 above)
Comment (pkirch) [reply .github/workflows/deploy.yaml:474]: The current addition follows the pattern of the steps in lines 419, 428, 437, 446, 451, 456, 471 which is used to call identity-hub-cli and registration-service-cli because the default working directory for the CLI is set in line 144.

```yaml
    defaults:
      run:
        working-directory: deployment/terraform/dataspace
```

However, the pattern `working-directory: .` is only used twice.

@algattik, therefore, I would not change. Makes sense?

File: .github/workflows/deploy.yaml (hunk 10 above)
Comment (pkirch) [reply .github/workflows/deploy.yaml]: Agreed! Will change.

File: README.md (hunk 12 above)
Comment (pkirch) [reply README.md]: Good approach! Thanks for sharing! Will change all occurrences.

File: system-tests/README.md (hunk 14 above)
Comment (pkirch) [reply system-tests/README.md]: Agreed! Thanks for bringing this up!

File: system-tests/README.md (hunk 15 above)
Comment (pkirch) [reply system-tests/README.md:196]: Agreed! Thanks for sharing!

File: system-tests/resources/cli-tools/entrypoint.sh (hunk 16 above)
Comment (pkirch) [reply system-tests/resources/cli-tools/entrypoint.sh:42]: The empty space belongs to the pattern and cannot be removed.

File: system-tests/resources/cli-tools/entrypoint.sh (hunk 17 above)
Comment (pkirch) [reply system-tests/resources/cli-tools/entrypoint.sh:38]: imho, this should be addressed separately.

This PR does not add or change behavior regarding the participants list (`PARTICIPANTS=(company1:eu company2:eu company3:us)`).
